core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -52,6 +57,288 @@ static TEE_Result ffa_get_dst(struct thread_smc_args *args,
+	 * once as a receiver in the memory share. For each receiver of a
+	 * memory share, we have one sp_mem_access_descr object.
+	 * This means that there can only be one SP linked to a specific
+	 * struct sp_mem_access_descr.
+	 */
+	SLIST_FOREACH(receiver, &smem->receivers, link) {
+		if (receiver->perm.endpoint_id == s->endpoint_id)
+			return receiver;
+	}
+	return NULL;
+}
+
+static TEE_Result add_mem_region_to_sp(struct ffa_mem_access *mem_acc,
+				       struct sp_mem *smem)
+{
+	struct ffa_mem_access_perm *access_perm =
+		&mem_acc->access_perm;
+	uint16_t endpoint_id = 0;
+	struct sp_session *s = NULL;
+	struct sp_mem_receiver *receiver = NULL;
+
+	endpoint_id = READ_ONCE(access_perm->endpoint_id);
+	s = sp_get_session(endpoint_id);
+
+	/* Only add memory shares of loaded SPs */
+	if (!s)
+		return FFA_DENIED;
+
+	/* Only allow each endpoint once */
+	if (get_sp_mem_receiver(s, smem))
+		return FFA_DENIED;
+
+	receiver = calloc(1, sizeof(struct sp_mem_receiver));
+
+	if (!receiver)
+		return FFA_NO_MEMORY;
+
+	receiver->smem = smem;
+
+	receiver->ref_count = 0;
+	memcpy(&receiver->perm, access_perm, sizeof(receiver->perm));
+
+	SLIST_INSERT_HEAD(&smem->receivers, receiver, link);
+
+	return FFA_OK;
+}
+
+static TEE_Result spmc_sp_handle_mem_share(struct thread_smc_args *args,
+					   struct ffa_rxtx *rxtx,
+					   struct sp_session *owner_sp)
+{
+	uint64_t global_handle = 0;
+	TEE_Result res = FFA_OK;
+	uint32_t ret_w2 = 0;
+	uint32_t ret_w3 = 0;
+
+	cpu_spin_lock(&rxtx->spinlock);
+
+	res = spmc_sp_add_share((struct ffa_mem_transaction *)rxtx->rx,
+				args->a1, &global_handle, owner_sp);
+	if (!res) {
+		reg_pair_from_64(global_handle, &ret_w3, &ret_w2);
+		args->a3 = ret_w3;
+		args->a2 = ret_w2;
+		args->a0 = FFA_SUCCESS_32;
+	} else {
+		ffa_set_error(args, res);
+	}
+
+	cpu_spin_unlock(&rxtx->spinlock);
+	return res;
+}
+
+static TEE_Result spmc_sp_add_sp_region(struct sp_mem *smem,
+					struct ffa_address_range *mem_reg,
+					struct sp_session *owner_sp,
+					uint8_t highest_permission)
+{
+	struct sp_ctx *sp_ctx = NULL;
+	uint64_t va = mem_reg->address;
+	uint64_t region_len = mem_reg->page_count * SMALL_PAGE_SIZE;
+
+	sp_ctx = to_sp_ctx(owner_sp->ts_sess.ctx);
+
+	/*
+	 * The memory region we try to share might not be linked to just 1
+	 * mobj. Create a new region for each mobj.
+	 */
+	while (region_len) {
+		uint64_t len = region_len;
+		struct sp_mem_map_region *region = NULL;
+		struct mobj *mobj = NULL;
+		uint16_t prot = 0;
+		uint64_t offs = 0;
+
+		/*
+		 * There is already a mobj for each address that is in the SPs
+		 * address range.
+		 */
+		mobj = vm_get_mobj(&sp_ctx->uctx, va, &len,
+				   &prot, &offs);
+
+		if (!mobj)
+			return FFA_DENIED;
+
+		/*
+		 * If we share memory from a SP, check if we are not sharing
+		 * with a higher permission than the memory was originally
+		 * mapped.
+		 */
+		if ((highest_permission & FFA_MEM_ACC_RW) &&
+		    !(prot & TEE_MATTR_PW)) {
+			return FFA_DENIED;
```

Comment:
Should call `mobj_put(mobj)` here and at the other places where `mobj` isn't saved.